### PR TITLE
New VehicleAction in MissionItem

### PR DESCRIPTION
**Pull request by rligocki**
based on:
- https://github.com/mavlink/MAVSDK-Proto/pull/226
- https://github.com/mavlink/MAVSDK-Swift/issues/198

### CODE REVIEW

File: src/plugins/mission/mission_impl.cpp
Comment (JonasVautherin) [inline src/plugins/mission/mission_impl.cpp:469]:
```suggestion
```

**Review by JonasVautherin — CHANGES_REQUESTED** [review]
Except for a wrong comment, it looks good to me :blush:. Did you test it? Does it work? :rocket:

**Comment by JonasVautherin** [comment]
Seeing the errors in CI (e.g. in `mission.h`), I come to wonder: did you auto-generate the code from the proto file, or did you write it manually? `mission.h` should be auto-generated, you should not have to write anything there yourself, right?

**Comment by julianoes** [comment]
Thank @rligocki. Now it would be nice if there was a new integration test using it.

**Comment by rligocki** [comment]
Yep I wrote it manually. Didn't know that it is possible to generate templates based on MAVSDK-Proto. I will have a look on generation of on C++ code based on Proto.

**Comment by julianoes** [comment]
@rligocki there is a comment telling you that the file is auto-generated.
And please have a look at the docs: https://mavsdk.mavlink.io/main/en/cpp/contributing/plugins.html#generate-h-and-cpp-files

**Comment by rligocki** [comment]
Now it looks like it might be ready for testing. What is best way to do it?

**Comment by JonasVautherin** [comment]
> What is best way to do it?

You can write a small script that uses it, or directly go for an integration test (like one of the `mission_*.cpp` [here](https://github.com/mavlink/MAVSDK/tree/main/src/integration_tests)). Maybe `mission_action.cpp`?

Also the style is wrong, please run `./tools/fix_style.sh .`, and the "proto" CI test is failing because you did not update the proto submodule in this PR. Just make it point to your MAVSDK-Proto PR commit and it should be fine :+1:.

**Comment by rligocki** [comment]
It looks like, there is some change in MAVSDK-Proto in core.proto file. Connection state is missing value 1. In my MAVSDK-Proto PR there are only additional code with VehicleAction. MAVSDK PR there is one removed line with that value.

**Comment by JonasVautherin** [comment]
> Connection state is missing value 1.

Wasn't this resolved in https://github.com/mavlink/MAVSDK/pull/1420? Is your branch up to date?

Also you forgot to fix the style: `./tools/fix_style.sh .`

**Comment by rligocki** [comment]
> > Connection state is missing value 1.
> 
> Wasn't this resolved in #1420? Is your branch up to date?
> 

https://github.com/mavlink/MAVSDK-Proto/blob/main/protos/core/core.proto - It looks like it is in main, so I was up to date.

> Also you forgot to fix the style: `./tools/fix_style.sh .`
> 

On my mac I was not able to compile these tools. So I need to do fix_style on different computer at home.

**Comment by JonasVautherin** [comment]
> It looks like it is in main, so I was up to date.

There two of them: the proto submodules, and then the MAVSDK branch itself. Both need to be up to date :+1:. This said, the CI tests pass... where do you see the issue with the connection state? When building the integration tests? If yes, just update your branch to `mavsdk:main` :+1:

**Comment by rligocki** [comment]
Still don't understand that problem with MAVSDK-Proto. On main branch, there is removed one connection state. Fixed problem with style. Should I create PR for connection state error?

**Comment by JonasVautherin** [comment]
> On main branch, there is removed one connection state. Fixed problem with style. Should I create PR for connection state error?

You have a fork of `mavlink/MAVSDK`, say `~/mavsdk`, right? Inside of it, there is the proto submodule, `~/mavsdk/proto`. Those are two different git repositories, but they are related.

You need `~/mavsdk/proto` to point to the HEAD of `mavlink/MAVSDK-Proto`, and `~/mavsdk` to point to the HEAD of `mavlink/MAVSDK`. The connection changes are both in the proto submodule (where the interface removes it) and in the mavsdk repo (where it is removed in the implementation)

Does that make sense?

**Comment by rligocki** [comment]
> You have a fork of `mavlink/MAVSDK`, say `~/mavsdk`, right? Inside of it, there is the proto submodule, `~/mavsdk/proto`. Those are two different git repositories, but they are related.
> 
This is what I understand. I see no problem there.

> You need `~/mavsdk/proto` to point to the HEAD of `mavlink/MAVSDK-Proto`, and `~/mavsdk` to point to the HEAD of `mavlink/MAVSDK`. The connection changes are both in the proto submodule (where the interface removes it) and in the mavsdk repo (where it is removed in the implementation)
But said that `~/mavsdk/proto` should point to MAVSDK-Proto PR. So you suggest to change proto git to point at mavlink/MAVSDK-Proto main branch?

**Comment by rligocki** [comment]
I am really lost. I need help at this situation. Cannot find where exactly I made mistake and how to solve it. Tried to find solution, but unfortunately it is first time I contribute to public repository so I have no experience at this case.

**Comment by JonasVautherin** [comment]
No worries, it can be tricky in the beginning :wink:. 

Should be fine now, both your branches were not up-to-date with upstream. Read about syncing with upstream e.g. [here](https://ardalis.com/syncing-a-fork-of-a-github-repository-with-upstream/) (it has to be done on both your mavsdk and mavsdk-proto forks until your proto is merged upstream).

**Comment by JonasVautherin** [comment]
The docs CI check is failing, I commented on the proto PR accordingly: https://github.com/mavlink/MAVSDK-Proto/pull/226

Other than that, it would be nice if you could add an integration test. Note that the integration tests run on the iris SITL multicopter, so ideally use actions that work for a multicopter (i.e. takeoff and land), and maybe write a test that checks that the iris behaves correctly when you use VTOL actions (I don't know what PX4 is expected to do, but a test for it would be nice).

I won't be asking you to change the integration tests such that they can run a VTOL, but that would be a nice addition later :blush:.

**Comment by rligocki** [comment]
Thanks a lot for your help. I though, that you would give me some advice, but this was faster way. 

Where are those integration tests located? I can check them and then create group of new tests for this purpose.

**Comment by julianoes** [comment]
Does this answer it? https://mavsdk.mavlink.io/main/en/cpp/guide/test.html

**Review by JonasVautherin — COMMENTED** [review]
Does the integration test prove that the takeoff action works? My feeling is that it would also takeoff without the takeoff action :thinking:.

Could we make it such that it lands and takes off in the middle of a mission? That I am pretty sure is not yet working :laughing:

File: src/integration_tests/mission_takeoff_land.cpp
Comment (rligocki) [inline src/integration_tests/mission_takeoff_land.cpp:81]: I am worried about &index. As I understand C++, &index is address of int value. Is it problem, then to do index++?

Comment (JonasVautherin) [reply src/integration_tests/mission_takeoff_land.cpp:81]: It's correct :+1:

Comment (JonasVautherin) [reply src/integration_tests/mission_takeoff_land.cpp:81]: Note that instead of using arrays, you could have used vectors. But I believe this works.

File: src/integration_tests/mission_transition.cpp
Comment (JonasVautherin) [inline src/integration_tests/mission_transition.cpp:87]: What's the expected behavior when flying a multicopter? I guess it is ignored, isn't it?

Comment (rligocki) [reply src/integration_tests/mission_transition.cpp:87]: I guess so. I was thinking, that if landing and takeoff would work, this action will provide that information.

File: src/integration_tests/mission_takeoff_land.cpp
Comment (rligocki) [reply src/integration_tests/mission_takeoff_land.cpp]: Modified PR to vectors. Thanks for your advices ;)

File: src/integration_tests/mission_takeoff_land.cpp
Comment (JonasVautherin) [inline src/integration_tests/mission_takeoff_land.cpp:168]: Either you want to do `landed_states_template.at(i)` and `landed_states.at(i)`, or remove the for loop (not sure if that works?), but something is wrong here :see_no_evil:

**Review by JonasVautherin — CHANGES_REQUESTED** [review]

File: src/integration_tests/mission_takeoff_land.cpp
Comment (rligocki) [reply src/integration_tests/mission_takeoff_land.cpp]: I just wanted to make sure, that those two vector are same. So I did check 8 times. 🙈🙈 My mistake. already fixed.

**Review by JonasVautherin — APPROVED** [review]

**Review by JonasVautherin — CHANGES_REQUESTED** [review]
LGTM, but will need to be merged after the acceptance_radius PR (because it requires an update of the proto submodule)

**Comment by JonasVautherin** [comment]
@rligocki: seems like the integration tests still fail. Would you mind having a look?

**Comment by rligocki** [comment]
I will try to look at it during today or tomorrow.

**Comment by JonasVautherin** [comment]
@rligocki: I merge the proto PR and update the branch here (don't forget to pull the changes). Once the integration tests are fixed, that can go in :+1:

**Comment by JonasVautherin** [comment]
@rligocki are you going to finish this or is it on hold for now? Asking because if it is on hold, I will just revert MAVSDK-Proto so that it does not block other features in the meantime :+1:

**Comment by rligocki** [comment]
Now I was able to found some time for this problem. I will fix those tests so that they will pass.

**Comment by julianoes** [comment]
@rligocki what's the status here? Would be nice to get this in soon.

**Comment by rligocki** [comment]
Cannot upload those vehicle mission item on PX4. There is an error.

**Comment by rligocki** [comment]
Currently there should be no error during runtime. One last thing, that needs to be done is transition mission validation. Now there is no measurement if transition happened. Also I cannot find way to run transition test with gazebo_standard_vtol plane.

**Pull request closed without merging**
